libvast/vast/concept/parseable/vast/schema.hpp

```diff
@@ -18,53 +18,183 @@
+      if (global != parent.global_symbols.end())
+        return global->second;
+      return caf::make_error(ec::parse_error, "undefined symbol:", key);
+    }
+
+    template <class Type>
+    caf::expected<type> operator()(Type x) {
+      return std::move(x);
+    }
+
+    caf::expected<type> operator()(const none_type& x) {
+      VAST_ASSERT(!x.name().empty());
+      auto concrete = lookup(x.name());
+      if (!concrete)
+        return concrete.error();
+      return concrete->update_attributes(x.attributes());
+    }
+
+    caf::expected<type> operator()(alias_type x) {
+      auto y = caf::visit(*this, x.value_type);
+      if (!y)
+        return y.error();
+      x.value_type = *y;
+      return std::move(x);
+    }
+
+    caf::expected<type> operator()(list_type x) {
+      auto y = caf::visit(*this, x.value_type);
+      if (!y)
+        return y.error();
+      x.value_type = *y;
+      return std::move(x);
+    }
+
+    caf::expected<type> operator()(map_type x) {
+      auto y = caf::visit(*this, x.value_type);
+      if (!y)
+        return y.error();
+      x.value_type = *y;
+      auto z = caf::visit(*this, x.key_type);
+      if (!z)
+        return z.error();
+      x.key_type = *z;
+      return std::move(x);
+    }
+
+    caf::expected<type> operator()(record_type x) {
+      for (auto& [field_name, field_type] : x.fields) {
+        auto y = caf::visit(*this, field_type);
+        if (!y)
+          return y.error();
+        field_type = *y;
+      }
+      return std::move(x);
+    }
+
+    caf::expected<type> resolve(symbol_buffer::iterator next) {
+      auto value = std::move(*next);
+      if (parent.local_symbols.find(value.first) != parent.local_symbols.end())
+        return caf::make_error(ec::parse_error, "duplicate definition of",
+                               value.first);
+      sb.erase(next);
+      auto x = caf::visit(*this, value.second);
+      if (!x)
+        return x.error();
+      auto [iter, inserted]
+        = parent.local_symbols.emplace(value.first, std::move(*x));
+      if (!inserted)
+        return caf::make_error(ec::parse_error, "failed to extend local "
+                                                "symbols");
+      auto added = sch.add(iter->second);
+      if (!added)
+        return caf::make_error(ec::parse_error, "failed to insert type",
+                               value.first);
+      return iter->second;
+    }
+
+    caf::expected<schema> run() {
+      while (!sb.empty()) {
+        if (auto x = resolve(sb.begin()); !x)
+          return x.error();
+      }
+      return sch;
+    }
+
+    const shared_schema_parser& parent;
+    symbol_buffer sb;
+    schema sch = {};
+  };
+
+  template <class Iterator, class Attribute>
+  bool parse(Iterator& f, const Iterator& l, Attribute& sch) const {
+    static_assert(detail::is_any_v<Attribute, attribute, unused_type>);
+    symbol_buffer sb;
+    bool table_failed = false;
```

Comment:
```suggestion
    bool duplicate_symbol = false;
```

Sounds like these are the semantics?